src/dsa/rpo_stark/signature/mod.rs

```diff
@@ -0,0 +1,176 @@
+use rand::{distributions::Uniform, prelude::Distribution, Rng};
+use winter_air::{FieldExtension, ProofOptions};
+use winter_math::{fields::f64::BaseElement, FieldElement};
+use winter_prover::Proof;
+use winter_utils::{ByteReader, ByteWriter, Deserializable, DeserializationError, Serializable};
+
+use crate::{
+    dsa::rpo_stark::stark::RpoSignatureScheme,
+    hash::{rpo::Rpo256, DIGEST_SIZE},
+    Word,
+};
+#[cfg(feature = "std")]
+use crate::{StarkField, ZERO};
+
+// CONSTANTS
+// ================================================================================================
+
+/// Specifies the parameters of the STARK underlying the signature scheme. These parameters provide
+/// at least 102 bits of security under the conjectured security of the toy protocol in
+/// the ethSTARK paper [1].
+///
+/// [1]: https://eprint.iacr.org/2021/582
+pub const PROOF_OPTIONS: ProofOptions =
+    ProofOptions::new(30, 8, 12, FieldExtension::Quadratic, 4, 7, true);
+
+// PUBLIC KEY
+// ================================================================================================
+
+/// A public key for verifying signatures.
+///
+/// The public key is a [Word] (i.e., 4 field elements) that is the hash of the secret key.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct PublicKey(Word);
+
+impl PublicKey {
+    /// Returns the [Word] defining the public key.
+    pub fn inner(&self) -> Word {
+        self.0
+    }
+}
+
+impl PublicKey {
+    /// Verifies the provided signature against provided message and this public key.
+    pub fn verify(&self, message: Word, signature: &Signature) -> bool {
+        signature.verify(message, *self)
+    }
+}
+
+impl Serializable for PublicKey {
+    fn write_into<W: ByteWriter>(&self, target: &mut W) {
+        self.0.write_into(target);
+    }
+}
+
+impl Deserializable for PublicKey {
+    fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError> {
+        let pk = <Word>::read_from(source)?;
+        Ok(Self(pk))
+    }
+}
+
+// SECRET KEY
+// ================================================================================================
+
+/// A secret key for generating signatures.
+///
+/// The secret key is a [Word] (i.e., 4 field elements).
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct SecretKey(Word);
+
+impl SecretKey {
+    /// Generates a secret key from OS-provided randomness.
+    pub fn new(word: Word) -> Self {
+        Self(word)
+    }
+
+    /// Generates a secret key from a [Word].
+    #[cfg(feature = "std")]
+    pub fn random() -> Self {
+        use rand::{rngs::StdRng, SeedableRng};
+
+        let mut rng = StdRng::from_entropy();
+        Self::with_rng(&mut rng)
+    }
+
+    /// Generates a secret_key using the provided random number generator `Rng`.
+    #[cfg(feature = "std")]
+    pub fn with_rng<R: Rng>(rng: &mut R) -> Self {
```

Comment:
Why do we need to gate this under `std`? For the Falcon signature we are able to avoid this.

Reply:
Indeed, changed to mimic the Falcon one